File: errorHandling/scala/src/main/scala/MonadThrow.scala
package ErrorHandling.MonadThrow

import cats.*
import cats.implicits.*
import cats.data.*
import cats.effect.*
import cats.effect.implicits.*
import UserThrowable.*

final case class User(
    id: Int,
    age: Int,
    name: String,
    address: String
)

enum UserThrowable extends Throwable:
  case NotFound
  case AlreadyInDB
  case InvalidAge
  case InvalidName

final case class UserRepository[F[_]: MonadThrow]():
  def add(
      age: Int,
      name: String,
      address: String
  ): F[Int] =
    if name == "Janos"
    then AlreadyInDB.raiseError
    else 1.pure

trait UserService[F[_]]:
  def createUser(
      age: Int,
      name: String,
      address: String
  ): F[User]

object UserService:
  def mkUserService[F[_]: MonadThrow](
      userRepository: UserRepository[F]
  ): UserService[F] = new UserService[F]:
    def createUser(
        age: Int,
        name: String,
        address: String
    ): F[User] =
      for {
        _ <- validateAge(age)
        _ <- validateName(name)
        id <- userRepository.add(age, name, address)
      } yield User(id, age, name, address)

    private def validateAge(
        age: Int
    ): F[Unit] =
      if age > 100 || age < 0
      then InvalidAge.raiseError
      else ().pure

    private def validateName(
        name: String
    ): F[Unit] =
      if name.size > 10
      then InvalidName.raiseError
      else ().pure

type EitherThrow[A] = Either[Throwable, A]

val service =
  UserService.mkUserService[EitherThrow](UserRepository())

// Problem: as IO has Throwable hard-coded into it,
// you cannot have an instance with your own error type

def makeUser(age: Int, name: String, address: String) =
  service
    .createUser(age, name, address)
    .handleError {
      case InvalidName => User(1, 2, "3", "4")
    }
    // .onError {
    //   case InvalidName => Right(println("invalid name"))
    //   case InvalidAge  => Right(println("invalid age"))
    // }

File: errorHandling/haskell/src/MonadThrow.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleContexts #-}

module MonadThrow where

import Control.Monad
import Control.Monad.Catch

data User = User Int Int String String deriving (Show, Eq)

data UserErr
  = NotFound
  | AlreadyInDB
  | InvalidAge
  | InvalidName
  deriving (Show, Eq)

instance Exception UserErr

newtype UserRepository m = UserRepository
  {add :: Int -> String -> String -> m Int}

mkUserRepoService :: (MonadThrow m) => UserRepository m
mkUserRepoService =
  UserRepository
    { add = \_ name _ ->
        if name == "John"
          then throwM AlreadyInDB
          else pure 1
    }

newtype UserService m = UserService
  {createUser :: Int -> String -> String -> m User}

mkUserService :: (MonadThrow m, MonadCatch m) => UserRepository m -> UserService m
mkUserService repo =
  UserService
    { createUser = \age name address -> do
        validateAge age
        validateName name
        uid <- add repo age name address
        return $ User uid age name address
    }

validateAge :: (MonadThrow m) => Int -> m ()
validateAge age =
  when (age > 100 || age < 0) (throwM InvalidAge)

validateName :: (MonadThrow m) => String -> m ()
validateName name =
  when (length name > 10) (throwM InvalidName)

service :: UserService IO
service = mkUserService mkUserRepoService

handler :: UserErr -> IO ()
handler = print

mkUser :: UserService m -> Int -> String -> String -> m User
mkUser = createUser

mkUser' :: UserService IO -> Int -> String -> String -> IO ()
mkUser' userService age name address = catch (createUser userService age name address >>= print) handler
